File: src/copilot_spend/paths.py
from __future__ import annotations

import json
import os
import stat
import tempfile
from pathlib import Path
from typing import Any

from .auth import AuthError


def config_dir() -> Path:
    override = os.environ.get("COPILOT_SPEND_CONFIG_DIR")
    if override:
        return Path(override)
    xdg = os.environ.get("XDG_CONFIG_HOME")
    if xdg:
        return Path(xdg) / "copilot-spend"
    return Path.home() / ".config" / "copilot-spend"


def auth_path() -> Path:
    return config_dir() / "auth.json"


def session_path() -> Path:
    return config_dir() / "session.json"


def assert_safe_parent(parent: Path) -> None:
    if os.name != "posix":
        return
    if not parent.exists():
        return
    try:
        info = parent.stat()
    except OSError as exc:
        raise AuthError(f"Cannot stat config directory {parent}: {exc}") from None
    if info.st_uid != os.getuid():
        raise AuthError(
            f"Refusing to use config directory {parent}: owned by uid "
            f"{info.st_uid}, expected {os.getuid()}. "
            "Set COPILOT_SPEND_CONFIG_DIR or XDG_CONFIG_HOME to a directory "
            "you own."
        )
    if info.st_mode & (stat.S_IWGRP | stat.S_IWOTH):
        raise AuthError(
            f"Refusing to use config directory {parent}: mode "
            f"{oct(stat.S_IMODE(info.st_mode))} grants group or world write. "
            f"Run `chmod 700 {parent}`."
        )


def write_secret_file(path: Path, payload: dict[str, Any]) -> None:
    parent = path.parent
    assert_safe_parent(parent)
    parent.mkdir(parents=True, exist_ok=True)
    if os.name == "posix":
        os.chmod(parent, 0o700)

    fd, tmp_str = tempfile.mkstemp(dir=parent, prefix=".tmp-", suffix=".json")
    tmp_path = Path(tmp_str)
    try:
        if os.name == "posix":
            os.chmod(fd, 0o600)
        with os.fdopen(fd, "wb") as handle:
            handle.write(json.dumps(payload, indent=2).encode("utf-8"))
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(tmp_path, path)
    except BaseException:
        tmp_path.unlink(missing_ok=True)
        raise


def delete_secret_file(path: Path) -> None:
    path.unlink(missing_ok=True)


def scrub(text: str, *tokens: str | None) -> str:
    for token in tokens:
        if token and token in text:
            text = text.replace(token, "<redacted-token>")
    return text

File: typescript/src/paths.ts
import { randomBytes } from "node:crypto";
import {
  chmodSync,
  closeSync,
  existsSync,
  fsyncSync,
  mkdirSync,
  openSync,
  renameSync,
  rmSync,
  statSync,
  writeSync,
} from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { AuthError } from "./errors";

const isPosix = process.platform !== "win32";

// stat.S_IWGRP | stat.S_IWOTH — group-write and world-write bits.
const GROUP_OR_WORLD_WRITE = 0o022;

export function configDir(): string {
  const override = process.env.COPILOT_SPEND_CONFIG_DIR;
  if (override) {
    return override;
  }
  const xdg = process.env.XDG_CONFIG_HOME;
  if (xdg) {
    return join(xdg, "copilot-spend");
  }
  return join(homedir(), ".config", "copilot-spend");
}

export function authPath(): string {
  return join(configDir(), "auth.json");
}

function octalMode(mode: number): string {
  return `0o${(mode & 0o7777).toString(8)}`;
}

export function assertSafeParent(parent: string): void {
  if (!isPosix) {
    return;
  }
  if (!existsSync(parent)) {
    return;
  }

  let info: ReturnType<typeof statSync>;
  try {
    info = statSync(parent);
  } catch (exc) {
    const reason = exc instanceof Error ? exc.message : String(exc);
    throw new AuthError(`Cannot stat config directory ${parent}: ${reason}`);
  }

  const uid = process.getuid?.();
  if (uid !== undefined && info.uid !== uid) {
    throw new AuthError(
      `Refusing to use config directory ${parent}: owned by uid ${info.uid}, expected ${uid}. Set COPILOT_SPEND_CONFIG_DIR or XDG_CONFIG_HOME to a directory you own.`,
    );
  }

  if (info.mode & GROUP_OR_WORLD_WRITE) {
    throw new AuthError(
      `Refusing to use config directory ${parent}: mode ${octalMode(info.mode)} grants group or ` +
        `world write. Run \`chmod 700 ${parent}\`.`,
    );
  }
}

export function writeSecretFile(path: string, payload: unknown): void {
  const parent = dirname(path);
  assertSafeParent(parent);
  mkdirSync(parent, { recursive: true });
  if (isPosix) {
    chmodSync(parent, 0o700);
  }

  const tmp = join(parent, `.tmp-${process.pid}-${randomBytes(8).toString("hex")}.json`);
  let fd: number | undefined;
  try {
    fd = openSync(tmp, "wx", 0o600);
    if (isPosix) {
      chmodSync(tmp, 0o600);
    }
    writeSync(fd, JSON.stringify(payload, null, 2));
    fsyncSync(fd);
    closeSync(fd);
    fd = undefined;
    renameSync(tmp, path);
  } catch (exc) {
    if (fd !== undefined) {
      try {
        closeSync(fd);
      } catch {
        // best-effort close before cleanup
      }
    }
    rmSync(tmp, { force: true });
    throw exc;
  }
}

export function deleteSecretFile(path: string): void {
  rmSync(path, { force: true });
}

export function scrub(text: string, ...tokens: (string | null | undefined)[]): string {
  let result = text;
  for (const token of tokens) {
    if (token) {
      result = result.split(token).join("<redacted-token>");
    }
  }
  return result;
}
